lib/counters/derive/src/lib.rs

```diff
@@ -51,60 +52,89 @@ fn gen_count_impl(input: DeriveInput) -> Result<impl ToTokens, syn::Error> {
         ));
     };
     let variants = &data_enum.variants;
-    let mut state = CountGenerator::new(&input.ident, variants.len());
+    let mut state = CountGenerator::new(&input, variants.len());
 
     for variant in variants {
         state.add_variant(variant)?;
     }
 
-    Ok(state.generate(&input))
+    Ok(state.generate())
 }
 
 struct CountGenerator<'input> {
+    input: &'input DeriveInput,
     enum_name: &'input syn::Ident,
     field_defs: Vec<proc_macro2::TokenStream>,
     field_inits: Vec<proc_macro2::TokenStream>,
     variant_patterns: Vec<proc_macro2::TokenStream>,
+    needed_generics: HashSet<syn::Ident>,
+    all_generics: HashSet<syn::Ident>,
+    where_clause_types: HashSet<syn::Type>,
     any_skipped: bool,
 }
 
 impl<'input> CountGenerator<'input> {
-    fn new(enum_name: &'input syn::Ident, variants: usize) -> Self {
+    fn new(input: &'input DeriveInput, variants: usize) -> Self {
         Self {
-            enum_name,
+            enum_name: &input.ident,
+            input,
             field_defs: Vec::with_capacity(variants),
             field_inits: Vec::with_capacity(variants),
             variant_patterns: Vec::with_capacity(variants),
+            all_generics: input
+                .generics
+                .type_params()
+                .map(|p| p.ident.clone())
+                .collect(),
+            needed_generics: HashSet::new(),
+            where_clause_types: HashSet::new(),
             any_skipped: false,
         }
     }
 
-    fn generate(self, input: &DeriveInput) -> impl ToTokens {
+    fn generate(self) -> impl ToTokens {
         let Self {
+            input,
             enum_name,
             field_defs,
             field_inits,
             mut variant_patterns,
             any_skipped,
+            needed_generics,
+            where_clause_types,
+            ..
         } = self;
 
+        let generics = &input.generics;
+
         // If we skipped any variants, generate a catchall case.
         if any_skipped {
             variant_patterns.push(quote! { _ => {} });
         }
         let vis = &input.vis;
-
         let counts_ty = counts_ty(enum_name);
+        let where_clauses = where_clause_types
+            .iter()
+            .map(|ty| {
+                quote! { #ty: counters::Count }
+            })
+            .collect::<Vec<_>>();
+        // I'm not sure why we have to do this, but quote gets mad if it's not a vec...
```

Comment:
FWIW I've also run into this.

Reply:
Yeah, as far as I understand it, it just needs to be `IntoIterator`, but it seems like the `HashSet`'s `IntoIterator` implementation doesn't meet some required bound in `quote` somewhere...